<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Contamination estimation with decontX — decontX • celda</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Contamination estimation with decontX — decontX" />
<meta property="og:description" content="Identifies contamination from factors such as ambient RNA
in single cell genomic datasets." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">celda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="http://celda.camplab.net/">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/celda_pbmc3k.html">Celda - Analysis of PBMC3K</a>
    </li>
    <li>
      <a href="../articles/articles/decontX_pbmc4k.html">DecontX - Decontamination of PBMC4K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://github.com/campbio/celda">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/campbio/celda/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Contamination estimation with decontX</h1>
    <small class="dont-index">Source: <a href='https://github.com/campbio/celda/blob/master/R/decon.R'><code>R/decon.R</code></a></small>
    <div class="hidden name"><code>decontX.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Identifies contamination from factors such as ambient RNA
in single cell genomic datasets.</p>
    </div>

    <pre class="usage"><span class='fu'>decontX</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>decontX</span><span class='op'>(</span>
  <span class='va'>x</span>,
  assayName <span class='op'>=</span> <span class='st'>"counts"</span>,
  z <span class='op'>=</span> <span class='cn'>NULL</span>,
  batch <span class='op'>=</span> <span class='cn'>NULL</span>,
  background <span class='op'>=</span> <span class='cn'>NULL</span>,
  bgAssayName <span class='op'>=</span> <span class='cn'>NULL</span>,
  maxIter <span class='op'>=</span> <span class='fl'>500</span>,
  delta <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>10</span><span class='op'>)</span>,
  estimateDelta <span class='op'>=</span> <span class='cn'>TRUE</span>,
  convergence <span class='op'>=</span> <span class='fl'>0.001</span>,
  iterLogLik <span class='op'>=</span> <span class='fl'>10</span>,
  varGenes <span class='op'>=</span> <span class='fl'>5000</span>,
  dbscanEps <span class='op'>=</span> <span class='fl'>1</span>,
  seed <span class='op'>=</span> <span class='fl'>12345</span>,
  logfile <span class='op'>=</span> <span class='cn'>NULL</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>

<span class='co'># S4 method for ANY</span>
<span class='fu'>decontX</span><span class='op'>(</span>
  <span class='va'>x</span>,
  z <span class='op'>=</span> <span class='cn'>NULL</span>,
  batch <span class='op'>=</span> <span class='cn'>NULL</span>,
  background <span class='op'>=</span> <span class='cn'>NULL</span>,
  maxIter <span class='op'>=</span> <span class='fl'>500</span>,
  delta <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>10</span><span class='op'>)</span>,
  estimateDelta <span class='op'>=</span> <span class='cn'>TRUE</span>,
  convergence <span class='op'>=</span> <span class='fl'>0.001</span>,
  iterLogLik <span class='op'>=</span> <span class='fl'>10</span>,
  varGenes <span class='op'>=</span> <span class='fl'>5000</span>,
  dbscanEps <span class='op'>=</span> <span class='fl'>1</span>,
  seed <span class='op'>=</span> <span class='fl'>12345</span>,
  logfile <span class='op'>=</span> <span class='cn'>NULL</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>A numeric matrix of counts or a SingleCellExperiment
with the matrix located in the assay slot under <code>assayName</code>.
Cells in each batch will be subsetted and converted to a sparse matrix
of class <code>dgCMatrix</code> from package Matrix before analysis. This
object should only contain filtered cells after cell calling. Empty
cell barcodes (low expression droplets before cell calling) are not needed
to run DecontX.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>For the generic, further arguments to pass to each method.</p></td>
    </tr>
    <tr>
      <th>assayName</th>
      <td><p>Character. Name of the assay to use if <code>x</code> is a
SingleCellExperiment.</p></td>
    </tr>
    <tr>
      <th>z</th>
      <td><p>Numeric or character vector. Cell cluster labels. If NULL,
PCA will be used to reduce the dimensionality of the dataset initially,
'<a href='https://rdrr.io/pkg/uwot/man/umap.html'>umap</a>' from the 'uwot' package
will be used to further reduce the dataset to 2 dimenions and
the '<a href='https://rdrr.io/pkg/dbscan/man/dbscan.html'>dbscan</a>' function from the 'dbscan' package
will be used to identify clusters of broad cell types. Default NULL.</p></td>
    </tr>
    <tr>
      <th>batch</th>
      <td><p>Numeric or character vector. Batch labels for cells.
If batch labels are supplied, DecontX is run on cells from each
batch separately. Cells run in different channels or assays
should be considered different batches. Default NULL.</p></td>
    </tr>
    <tr>
      <th>background</th>
      <td><p>A numeric matrix of counts or a
SingleCellExperiment with the matrix located in the assay
slot under <code>assayName</code>. It should have the same structure as <code>x</code>
except it contains the matrix of empty droplets instead of cells. When
supplied, empirical distribution of transcripts from these empty droplets
will be used as the contamination distribution. Default NULL.</p></td>
    </tr>
    <tr>
      <th>bgAssayName</th>
      <td><p>Character. Name of the assay to use if <code>background</code>
is a SingleCellExperiment. Default to same as
<code>assayName</code>.</p></td>
    </tr>
    <tr>
      <th>maxIter</th>
      <td><p>Integer. Maximum iterations of the EM algorithm. Default 500.</p></td>
    </tr>
    <tr>
      <th>delta</th>
      <td><p>Numeric Vector of length 2. Concentration parameters for
the Dirichlet prior for the contamination in each cell. The first element
is the prior for the native counts while the second element is the prior for
the contamination counts. These essentially act as pseudocounts for the
native and contamination in each cell. If <code>estimateDelta = TRUE</code>,
this is only used to produce a random sample of proportions for an initial
value of contamination in each cell. Then
<code><a href='https://rdrr.io/pkg/MCMCprecision/man/fit_dirichlet.html'>fit_dirichlet</a></code> is used to update
<code>delta</code> in each iteration.
If <code>estimateDelta = FALSE</code>, then <code>delta</code> is fixed with these
values for the entire inference procedure. Fixing <code>delta</code> and
setting a high number in the second element will force <code>decontX</code>
to be more aggressive and estimate higher levels of contamination at
the expense of potentially removing native expression.
Default <code><a href='https://rdrr.io/r/base/c.html'>c(10, 10)</a></code>.</p></td>
    </tr>
    <tr>
      <th>estimateDelta</th>
      <td><p>Boolean. Whether to update <code>delta</code> at each
iteration.</p></td>
    </tr>
    <tr>
      <th>convergence</th>
      <td><p>Numeric. The EM algorithm will be stopped if the maximum
difference in the contamination estimates between the previous and
current iterations is less than this. Default 0.001.</p></td>
    </tr>
    <tr>
      <th>iterLogLik</th>
      <td><p>Integer. Calculate log likelihood every <code>iterLogLik</code>
iteration. Default 10.</p></td>
    </tr>
    <tr>
      <th>varGenes</th>
      <td><p>Integer. The number of variable genes to use in
dimensionality reduction before clustering. Variability is calcualted using
<code><a href='https://rdrr.io/pkg/scran/man/modelGeneVar.html'>modelGeneVar</a></code> function from the 'scran' package.
Used only when z is not provided. Default 5000.</p></td>
    </tr>
    <tr>
      <th>dbscanEps</th>
      <td><p>Numeric. The clustering resolution parameter
used in '<a href='https://rdrr.io/pkg/dbscan/man/dbscan.html'>dbscan</a>' to estimate broad cell clusters.
Used only when z is not provided. Default 1.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>Integer. Passed to <a href='https://withr.r-lib.org/reference/with_seed.html'>with_seed</a>. For reproducibility,
a default value of 12345 is used. If NULL, no calls to
<a href='https://withr.r-lib.org/reference/with_seed.html'>with_seed</a> are made.</p></td>
    </tr>
    <tr>
      <th>logfile</th>
      <td><p>Character. Messages will be redirected to a file named
`logfile`. If NULL, messages will be printed to stdout.  Default NULL.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical. Whether to print log messages. Default TRUE.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>If <code>x</code> is a matrix-like object, a list will be returned
with the following items:</p><dl>
<dt><code>decontXcounts</code>:</dt><dd><p>The decontaminated matrix. Values obtained
from the variational inference procedure may be non-integer. However,
integer counts can be obtained by rounding,
e.g. <code><a href='https://rdrr.io/r/base/Round.html'>round(decontXcounts)</a></code>.</p></dd>
<dt><code>contamination</code>:</dt><dd><p>Percentage of contamination in each cell.</p></dd>
<dt><code>estimates</code>:</dt><dd><p>List of estimated parameters for each batch. If z
was not supplied, then the UMAP coordinates used to generated cell
cluster labels will also be stored here.</p></dd>
<dt><code>z</code>:</dt><dd><p>Cell population/cluster labels used for analysis.</p></dd>
<dt><code>runParams</code>:</dt><dd><p>List of arguments used in the function call.</p></dd>

</dl>

<p>If <code>x</code> is a <a href='https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html'>SingleCellExperiment</a>, then the decontaminated
counts will be stored as an assay and can be accessed with
<code><a href='decontXcounts.html'>decontXcounts(x)</a></code>. The contamination values and cluster labels
will be stored in <code>colData(x)</code>. <code>estimates</code> and <code>runParams</code>
will be stored in <code>metadata(x)$decontX</code>. The UMAPs used to generated
cell cluster labels will be stored in
<code>reducedDims</code> slot in <code>x</code>.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Shiyi Yang, Yuan Yin, Joshua Campbell</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Generate matrix with contamination</span>
<span class='va'>s</span> <span class='op'>&lt;-</span> <span class='fu'><a href='simulateContamination.html'>simulateContamination</a></span><span class='op'>(</span>seed <span class='op'>=</span> <span class='fl'>12345</span><span class='op'>)</span>

<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>SingleCellExperiment</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: SummarizedExperiment</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: MatrixGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: matrixStats</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘MatrixGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:matrixStats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
#&gt; <span class='message'>    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
#&gt; <span class='message'>    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
#&gt; <span class='message'>    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
#&gt; <span class='message'>    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
#&gt; <span class='message'>    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
#&gt; <span class='message'>    colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
#&gt; <span class='message'>    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
#&gt; <span class='message'>    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
#&gt; <span class='message'>    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
#&gt; <span class='message'>    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
#&gt; <span class='message'>    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
#&gt; <span class='message'>    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
#&gt; <span class='message'>    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
#&gt; <span class='message'>    rowWeightedSds, rowWeightedVars</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BiocGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:parallel’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
#&gt; <span class='message'>    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
#&gt; <span class='message'>    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
#&gt; <span class='message'>    union, unique, unsplit, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘S4Vectors’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:celda’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    params</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘Biobase’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:MatrixGenerics’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    rowMedians</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:matrixStats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyMissing, rowMedians</span></div><div class='input'><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html'>SingleCellExperiment</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>s</span><span class='op'>$</span><span class='va'>observedCounts</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>decontX</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>--------------------------------------------------</span></div><div class='output co'>#&gt; <span class='message'>Starting DecontX</span></div><div class='output co'>#&gt; <span class='message'>--------------------------------------------------</span></div><div class='output co'>#&gt; <span class='message'>Wed Sep 29 12:24:56 2021 .. Analyzing all cells</span></div><div class='output co'>#&gt; <span class='message'>Wed Sep 29 12:24:56 2021 .... Converting to sparse matrix</span></div><div class='output co'>#&gt; <span class='message'>Wed Sep 29 12:24:57 2021 .... Generating UMAP and estimating cell types</span></div><div class='output co'>#&gt; <span class='message'>Wed Sep 29 12:25:00 2021 .... Estimating contamination</span></div><div class='output co'>#&gt; <span class='message'>Wed Sep 29 12:25:00 2021 ...... Completed iteration: 9 | converge: 0.0009154</span></div><div class='output co'>#&gt; <span class='message'>--------------------------------------------------</span></div><div class='output co'>#&gt; <span class='message'>Completed DecontX. Total time: 3.44309 secs</span></div><div class='output co'>#&gt; <span class='message'>--------------------------------------------------</span></div><div class='input'>
<span class='co'># Plot contamination on UMAP</span>
<span class='fu'><a href='plotDecontXContamination.html'>plotDecontXContamination</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
</div><div class='img'><img src='decontX-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Plot decontX cluster labels</span>
<span class='va'>umap</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html'>reducedDim</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='fu'><a href='plotDimReduceCluster.html'>plotDimReduceCluster</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>decontX_clusters</span>,
    dim1 <span class='op'>=</span> <span class='va'>umap</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>]</span>, dim2 <span class='op'>=</span> <span class='va'>umap</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span>, <span class='op'>)</span>
</div><div class='img'><img src='decontX-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Plot percentage of marker genes detected</span>
<span class='co'># in each cell cluster before decontamination</span>
<span class='va'>s</span><span class='op'>$</span><span class='va'>markers</span>
</div><div class='output co'>#&gt; $CellType_1_Markers
#&gt; [1] "Gene_47" "Gene_32" "Gene_86"
#&gt; 
#&gt; $CellType_2_Markers
#&gt; [1] "Gene_70" "Gene_33" "Gene_48"
#&gt; 
#&gt; $CellType_3_Markers
#&gt; [1] "Gene_74" "Gene_26" "Gene_20"
#&gt; </div><div class='input'><span class='fu'><a href='plotDecontXMarkerPercentage.html'>plotDecontXMarkerPercentage</a></span><span class='op'>(</span><span class='va'>sce</span>, markers <span class='op'>=</span> <span class='va'>s</span><span class='op'>$</span><span class='va'>markers</span>, assayName <span class='op'>=</span> <span class='st'>"counts"</span><span class='op'>)</span>
</div><div class='img'><img src='decontX-3.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Plot percentage of marker genes detected</span>
<span class='co'># in each cell cluster after contamination</span>
<span class='fu'><a href='plotDecontXMarkerPercentage.html'>plotDecontXMarkerPercentage</a></span><span class='op'>(</span><span class='va'>sce</span>, markers <span class='op'>=</span> <span class='va'>s</span><span class='op'>$</span><span class='va'>markers</span>,
                            assayName <span class='op'>=</span> <span class='st'>"decontXcounts"</span><span class='op'>)</span>
</div><div class='img'><img src='decontX-4.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Plot percentage of marker genes detected in each cell</span>
<span class='co'># comparing original and decontaminated counts side-by-side</span>
<span class='fu'><a href='plotDecontXMarkerPercentage.html'>plotDecontXMarkerPercentage</a></span><span class='op'>(</span><span class='va'>sce</span>, markers <span class='op'>=</span> <span class='va'>s</span><span class='op'>$</span><span class='va'>markers</span>,
                            assayName <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"counts"</span>, <span class='st'>"decontXcounts"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='img'><img src='decontX-5.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Plot raw counts of indiviual markers genes before</span>
<span class='co'># and after decontamination</span>
<span class='fu'><a href='plotDecontXMarkerExpression.html'>plotDecontXMarkerExpression</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='va'>s</span><span class='op'>$</span><span class='va'>markers</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='img'><img src='decontX-6.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Joshua Campbell, Shiyi Yang, Zhe Wang, Sean Corbett, Yusuke Koga.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


